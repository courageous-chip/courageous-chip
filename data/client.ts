import { ApolloClient } from "@apollo/client";

import { cache } from "./cache";
import { link } from "./links/link";

export const client = new ApolloClient({ cache, link });
