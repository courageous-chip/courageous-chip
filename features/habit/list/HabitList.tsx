import { gql, useQuery } from "@apollo/client";
import { useTheme } from "@react-navigation/native";
import React, { FC } from "react";
import { FlatList, StyleSheet, View } from "react-native";

import { EmptyView } from "../../ui/EmptyView";
import { ErrorView } from "../../ui/ErrorView";
import { LoadingView } from "../../ui/LoadingView";
import { ThemedItemSeparator } from "../../ui/ThemedItemSeparator";
import {
  HabitListItemOnPress,
  HABIT_LIST_ITEM_FIELDS_FRAGMENT,
  keyExtractor,
  renderItem,
} from "./HabitListItem";
import { ListHabitsQuery } from "./__generated__/ListHabitsQuery";

type Props = { onPress: HabitListItemOnPress };

export const HabitList: FC<Props> = function ({ onPress }) {
  const {
    colors: { background: backgroundColor },
  } = useTheme();
  const { data, error, loading } = useQuery<ListHabitsQuery>(LIST_HABITS_QUERY);

  return loading ? (
    <LoadingView />
  ) : error ? (
    <ErrorView />
  ) : !data?.habits.length ? (
    <EmptyView />
  ) : (
    <View style={[styles.container, { backgroundColor }]}>
      <FlatList
        data={data.habits}
        ItemSeparatorComponent={ThemedItemSeparator}
        keyExtractor={keyExtractor}
        renderItem={renderItem(onPress)}
      />
    </View>
  );
};

export const LIST_HABITS_QUERY = gql`
  query ListHabitsQuery {
    habits {
      ...HabitListItemFields
    }
  }

  ${HABIT_LIST_ITEM_FIELDS_FRAGMENT}
`;

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
