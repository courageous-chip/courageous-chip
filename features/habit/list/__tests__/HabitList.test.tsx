import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { render, wait } from "@testing-library/react-native";
import { GraphQLError } from "graphql";
import React from "react";

import { HabitList, GET_HABITS_QUERY } from "../HabitList";

describe("<HabitList />", () => {
  it("renders as empty", async () => {
    const mocks: MockedResponse[] = [
      {
        request: { query: GET_HABITS_QUERY },
        result: {
          data: {
            habits: [],
          },
        },
      },
    ];
    const { baseElement, getByText } = render(
      <MockedProvider mocks={mocks}>
        <HabitList />
      </MockedProvider>,
    );

    await wait(() => expect(getByText("ü§∑‚Äç‚ôÄÔ∏è")));

    expect(baseElement).toMatchSnapshot();
  });

  it("renders habits", async () => {
    const mocks: MockedResponse[] = [
      {
        request: { query: GET_HABITS_QUERY },
        result: {
          data: {
            habits: [{ id: "1", name: "ü¶∑ Floss Teeth", __typename: "Habit" }],
          },
        },
      },
    ];
    const { baseElement, getByText } = render(
      <MockedProvider mocks={mocks}>
        <HabitList />
      </MockedProvider>,
    );

    await wait(() => expect(getByText("ü¶∑ Floss Teeth")));

    expect(baseElement).toMatchSnapshot();
  });

  it("renders as loading", async () => {
    const mocks: MockedResponse[] = [
      {
        request: { query: GET_HABITS_QUERY },
        result: {},
      },
    ];
    const { baseElement, getByText } = render(
      <MockedProvider mocks={mocks}>
        <HabitList />
      </MockedProvider>,
    );

    await wait(() => expect(getByText("ü§î")));

    expect(baseElement).toMatchSnapshot();
  });

  it("renders on error", async () => {
    const mocks: MockedResponse[] = [
      {
        request: { query: GET_HABITS_QUERY },
        result: {
          errors: [new GraphQLError("üôÖ‚Äç‚ôÄÔ∏è")],
        },
      },
    ];
    const { baseElement, getByText } = render(
      <MockedProvider mocks={mocks}>
        <HabitList />
      </MockedProvider>,
    );

    await wait(() => expect(getByText("üí©")));

    expect(baseElement).toMatchSnapshot();
  });
});
