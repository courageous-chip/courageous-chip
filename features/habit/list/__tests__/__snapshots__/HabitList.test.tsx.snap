// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HabitList /> renders as empty 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 72,
          }
        }
      >
        🤷‍♀️
      </Text>
    </View>
  </View>
</View>
`;

exports[`<HabitList /> renders as loading 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 72,
          }
        }
      >
        🤔
      </Text>
    </View>
  </View>
</View>
`;

exports[`<HabitList /> renders habits 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
        }
      }
    >
      <ScrollView
        data={
          Array [
            Object {
              "__typename": "Habit",
              "id": "1",
              "name": "🦷 Floss Teeth",
            },
          ]
        }
        disableVirtualization={false}
        horizontal={false}
        initialNumToRender={10}
        maxToRenderPerBatch={10}
        numColumns={1}
        onEndReachedThreshold={2}
        removeClippedSubviews={false}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View
          style={null}
        >
          <View
            style={
              Object {
                "justifyContent": "center",
                "minHeight": 64,
              }
            }
          >
            <Text
              style={
                Object {
                  "marginHorizontal": 20,
                }
              }
            >
              🦷 Floss Teeth
            </Text>
          </View>
        </View>
      </ScrollView>
    </View>
  </View>
</View>
`;

exports[`<HabitList /> renders on error 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 72,
          }
        }
      >
        💩
      </Text>
    </View>
  </View>
</View>
`;
